From 0a8578ced218a082553ff271970502f3aefbee2f Mon Sep 17 00:00:00 2001
From: Hua Wang <Hua.Wang@windriver.com>
Date: Tue, 8 Jan 2013 09:41:02 +0800
Subject: AAC: support AAC Main Profile

FDK-AAC doesn't support AAC main profile, when type is AOT_AAC_MAIN,
it shows AAC_DEC_UNSUPPORTED_FORMAT. So it need to add AOT_AAC_MAIN
and deal with  PredictorDataPresent(reference open source faad2-2.7
as for http://sourcecodebrowser.com/faad2/2.7/syntax_8c.html.)

Category: aosp improvement
Domain: AUDIO_MEDIA-External
Origin: internal
Upstream-Candidate: yes
Issue:AXIA-1374
Change-Id: I331a5a63e35b2afc2eea0fe470c64fe3304e5f4c
Signed-off-by: Hua Wang <Hua.Wang@windriver.com>
---
 libAACdec/src/aacdecoder.cpp   |    1 +
 libAACdec/src/channelinfo.cpp  |   12 ++++++++++--
 libFDK/src/FDK_tools_rom.cpp   |    1 +
 libMpegTPDec/src/tpdec_asc.cpp |    2 ++
 4 files changed, 14 insertions(+), 2 deletions(-)

diff --git a/libAACdec/src/aacdecoder.cpp b/libAACdec/src/aacdecoder.cpp
index 3a2a561..cbd5508 100644
--- a/libAACdec/src/aacdecoder.cpp
+++ b/libAACdec/src/aacdecoder.cpp
@@ -827,6 +827,7 @@ LINKSPEC_CPP AAC_DECODER_ERROR CAacDecoder_Init(HANDLE_AACDECODER self, const CS
     self->streamInfo.profile = 1;
     break;
 
+  case AOT_AAC_MAIN:
   case AOT_SBR:
   case AOT_PS:
   case AOT_ER_AAC_LD:
diff --git a/libAACdec/src/channelinfo.cpp b/libAACdec/src/channelinfo.cpp
index 76d5895..788793f 100644
--- a/libAACdec/src/channelinfo.cpp
+++ b/libAACdec/src/channelinfo.cpp
@@ -165,8 +165,16 @@ AAC_DECODER_ERROR IcsRead(HANDLE_FDK_BITSTREAM bs,
     {
       if ((UCHAR)FDKreadBits(bs,1) != 0 ) /* UCHAR PredictorDataPresent */
       {
-        ErrorStatus = AAC_DEC_UNSUPPORTED_PREDICTION;
-        goto bail;
+        int sfb;
+        int limit = (pIcsInfo->TotalSfBands < pIcsInfo->MaxSfBands) ? pIcsInfo->TotalSfBands : pIcsInfo->MaxSfBands;
+
+        if ((UCHAR)FDKreadBits(bs,1) != 0) {
+            FDKreadBits(bs,5);
+        }
+
+        for (sfb = 0; sfb < limit; sfb++) {
+            FDKreadBits(bs,1);
+        }
       }
     }
 
diff --git a/libFDK/src/FDK_tools_rom.cpp b/libFDK/src/FDK_tools_rom.cpp
index c6e517e..3586cb0 100644
--- a/libFDK/src/FDK_tools_rom.cpp
+++ b/libFDK/src/FDK_tools_rom.cpp
@@ -2388,6 +2388,7 @@ const element_list_t * getBitstreamElementList(AUDIO_OBJECT_TYPE aot, SCHAR epCo
 {
   switch (aot) {
     case AOT_AAC_LC:
+    case AOT_AAC_MAIN:
     case AOT_SBR:
     case AOT_PS:
       FDK_ASSERT(epConfig == -1);
diff --git a/libMpegTPDec/src/tpdec_asc.cpp b/libMpegTPDec/src/tpdec_asc.cpp
index 4d7cc02..dee616f 100644
--- a/libMpegTPDec/src/tpdec_asc.cpp
+++ b/libMpegTPDec/src/tpdec_asc.cpp
@@ -954,6 +954,7 @@ TRANSPORTDEC_ERROR AudioSpecificConfig_Parse(
   {
 #ifdef TP_GA_ENABLE
     case AOT_AAC_LC:
+    case AOT_AAC_MAIN:
     case AOT_ER_AAC_LC:
     case AOT_ER_AAC_LD:
     case AOT_ER_AAC_SCAL:
@@ -999,6 +1000,7 @@ TRANSPORTDEC_ERROR AudioSpecificConfig_Parse(
   {
 #if defined(TP_GA_ENABLE) || defined(TP_USAC_ENABLE)
     case AOT_AAC_LC:
+    case AOT_AAC_MAIN:
     case AOT_ER_AAC_LC:
     case AOT_ER_AAC_SCAL:
     case AOT_ER_BSAC:
-- 
1.7.4.1

